src/cargo/ops/cargo_install.rs

```diff
@@ -85,15 +566,38 @@ pub fn install(
                 no_track,
                 !did_update,
             ) {
-                Ok(still_needs_update) => {
+                Ok(Some(installable_pkg)) => {
+                    did_update = true;
+                    Some((krate, installable_pkg))
+                }
+                Ok(None) => {
+                    // Already installed
                     succeeded.push(krate);
-                    did_update |= !still_needs_update;
+                    None
                 }
                 Err(e) => {
                     crate::display_error(&e, &mut config.shell());
                     failed.push(krate);
                     // We assume an update was performed if we got an error.
                     did_update = true;
+                    None
+                }
+            }
+        });
```

Comment:
Just noticed that I actually have to `collect()` here, or else we won't successfully be completing all of the package selection prior to the compilation process. Once I add collect() here, I have to update a few tests, which is a good sanity check.